feat(progress): disable progress in cmdline via messagesopt'

shadmansaleh opened the pull request:
Problem:
Currently, there's no way to tell neovim not to show progress messages in message area. This would be useful because third-party plugins can listen to the Progress event and show the progress status in their own way. But currently, it'd lead to the progress message being shown twice. Once by the building message system and once by the plugin.

Solution:
Add progress: entry in existing 'messageopts' option to control this behavior

Adds progress:c to messageopts as default. Here c stands for cmd. Later we can add things like s for statusline and etc as we introduce them. Using one char basically allows for easy mixing and matching of different combinations of possible values.

closes #35535

luukvbaal reviewed (approved):
No objections. Yes I still think moving (some) UI2 options to `messagesopts` is the way forwards if it stabilizes. But as echasnovski mentions this seems forward compatible with the approach in #35535.

In case @clason was concerned about my comment regarding lumping all message kinds into a single option, I think that only would provide real benefit over the approach taken here if we had actual table options😭. As it stands the other kind/target pairs can just be recognized in `messagesopts` in the future.

shadmansaleh commented:
> I don't think UI2 would ever itself set progress: to then use the autocmd event, that seems quite backwards

Nah not like that. Say once ui2 aupports progress we can add another flag to `progress:` for it. For example, a floating progress bar in ui2. Users can use that flag to enable that ui2 feature.

Either way, for now current setup is fine. This can be iterated on once other stuff supports progress

shadmansaleh committed:
feat(progress): add option to hide progress messages from cmdline

Problem:
Currently, there's no way to tell neovim not to show progress messages
in message area. This would be useful because third-party plugins can
listen to the Progress event and show the progress status in their own
way. But currently, it'd lead to the progress message being shown twice.
Once by the building message system and once by the plugin.

Solution:
Add progress: entry in existing 'messageopts' option to control
this behavior.

luukvbaal reviewed (commented):
My stance on this hasn't changed since https://github.com/neovim/neovim/pull/36730#pullrequestreview-3519094319. The configuration interface added to the ui2 module in #38091 could eventually be parsed by 'messagesopt'.

clason commented:
> Not the biggest fan of the flag style option either but it's what we have.

I agree; I just wanted to raise the point that `ui2` presents the chance of a clean break _if we want to_. 

I guess the question is how flexible the routing will become: If we are happy with just three kinds and three targets, `mopt` is as good a place as any for that. But if we start adding custom handlers (e.g., native or GUI) or dynamic routing (i.e., the "target" is a function that dispatches conditionally) -- which I think we will want to eventually -- this becomes much trickier, and we would have to consider the full picture before we can design an interface.

I guess a simple "default" interface ("`mopt = 'progress:m,list:p,message:c'` is the default; if you don't like the floating `msg` target just set `progress:c`") plus a -- to be discussed -- lower level fully configurable Lua interface could work well?

The important bit is not to paint ourselves into a corner too early (with a partial solution added in a release that we'd end up deprecating soon after).

echasnovski commented:
> That would supposedly resolve the problem statement here.

Partially, maybe. The `nvim_echo` progress messages would still appear in the history (one per progress), which is a bit confusing but probably not too much.

My concern from a plugin author point of view is that there is an "official" way to change how progress report is displayed. This was *the* request from very early on. Right now it is not quite possible without hacks (like hiding the message), but if it is the best solution overall - well okay.

I personally think that 'messagesopt' is the right place to put the logic of how progress messages are displayed. It is described as "Option settings for outputting messages." after all.

luukvbaal commented on test/old/testdir/setup.vim:
```diff
@@ -18,6 +18,7 @@ if exists('s:did_load')
   set laststatus=1
   set listchars=eol:$
   set maxsearchcount=99
+  set messagesopt=hit-enter,history:500
```
Why was this added here? Removing it in #39280 so those messages are shown in the cmdline by default in oldtests.

shadmansaleh replied:
As far as I remmeber I was getting old test faliures is some unreleted modules after the patch. I had added it solve that. But seems those falures don't happen now so no issue removing this. Maybe some other patch already fixed it in some different manner.
